ipfx/data_set_features.py

```diff
@@ -162,66 +162,97 @@ def extract_cell_features(data_set,
-    lsq_features = lsq_an.analyze(lsq_sweeps)
-    cell_features["long_squares"] = lsq_an.as_dict(lsq_features, [ dict(sweep_number=sn) for sn in long_square_sweep_numbers ])
-
-    if cell_features["long_squares"]["hero_sweep"] is None:
-        raise er.FeatureError("Could not find hero sweep.")
+    try:
+        lu.log_pretty_header("Long Squares:",level=2)
+        if len(long_square_sweep_numbers) == 0:
+            raise er.FeatureError("No long_square sweeps available for feature extraction")
+
+        lsq_sweeps = data_set.sweep_set(long_square_sweep_numbers)
+        lsq_sweeps.select_epoch("recording")
+        lsq_sweeps.align_to_start_of_epoch("experiment")
+        # TODO: fail properly when stim sweeps differ in start/duration
+        lsq_start, lsq_dur, _, _, _ = stf.get_stim_characteristics(lsq_sweeps.sweeps[0].i, lsq_sweeps.sweeps[0].t)
+
+        lsq_spx, lsq_spfx = extractors_for_sweeps(lsq_sweeps,
+                                                start = lsq_start,
+                                                end = lsq_start+lsq_dur,
+                                                **detection_parameters(data_set.LONG_SQUARE))
+        lsq_an = spa.LongSquareAnalysis(lsq_spx, lsq_spfx, subthresh_min_amp=subthresh_min_amp, require_subthreshold=False, require_suprathreshold=False)
+        lsq_features = lsq_an.analyze(lsq_sweeps)
+        cell_features["long_squares"] = lsq_an.as_dict(lsq_features, [ dict(sweep_number=sn) for sn in long_square_sweep_numbers ])
+
+        if cell_features["long_squares"]["hero_sweep"] is None:
+            raise er.FeatureError("Could not find hero sweep.")
+
+    except (er.FeatureError,IndexError) as e:
+        cell_features["long_squares"] = None
+        logging.warning(e)
+        if cell_state["failed_fx"]:
+            cell_state["fail_fx_message"].append('; '+str(e))
+        else:
+            cell_state["failed_fx"] = True
+            cell_state["fail_fx_message"] = str(e)
 
     # short squares
-    lu.log_pretty_header("Short Squares:",level=2)
-
-    if len(short_square_sweep_numbers) == 0:
-        raise er.FeatureError("No short square sweeps available for feature extraction")
-
-    ssq_sweeps = data_set.sweep_set(short_square_sweep_numbers)
-    ssq_sweeps.select_epoch("recording")
-    ssq_sweeps.align_to_start_of_epoch("experiment")
-
-    ssq_start, ssq_dur, _, _, _ = stf.get_stim_characteristics(ssq_sweeps.sweeps[0].i, ssq_sweeps.sweeps[0].t)
-    ssq_spx, ssq_spfx = extractors_for_sweeps(ssq_sweeps,
-                                              est_window = [ssq_start,ssq_start+0.001],
-                                              **detection_parameters(data_set.SHORT_SQUARE))
-    ssq_an = spa.ShortSquareAnalysis(ssq_spx, ssq_spfx)
-    ssq_features = ssq_an.analyze(ssq_sweeps)
-    cell_features["short_squares"] = ssq_an.as_dict(ssq_features, [ dict(sweep_number=sn) for sn in short_square_sweep_numbers ])
+    try:
```

Comment:
These all seem extremely similar, particularly the error handling code, which looks to be exactly duplicated up to a single string. I think these common components ought to be refactored into reusable functions.